### PR TITLE
claim contract

### PR DESCRIPTION
**Pull request by jhernandezb**

### CODE REVIEW

File: contracts/claim/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use cw2::set_contract_version;
+use minter::msg::{MintCountResponse, QueryMsg};
+use sg_std::{create_claim_for_msg, StargazeMsgWrapper};
+
+use crate::error::ContractError;
+use crate::msg::{ExecuteMsg, InstantiateMsg};
+pub type Response = cosmwasm_std::Response<StargazeMsgWrapper>;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:claim";
```
Comment (shanev) [inline contracts/claim/src/contract.rs:14]:
```suggestion
const CONTRACT_NAME: &str = "crates.io:sg-claim";
```

File: contracts/claim/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    _msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    Ok(Response::new().add_attribute("method", "instantiate"))
```
Comment (shanev) [inline contracts/claim/src/contract.rs:25]:
```suggestion
    Ok(Response::new().add_attribute("action", "instantiate"))
```

File: contracts/claim/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::ClaimMintNFT { minter_address } => {
+            try_claim_mint_nft(deps, info.sender, minter_address)
+        }
+    }
+}
+
+pub fn try_claim_mint_nft(
```
Comment (shanev) [inline contracts/claim/src/contract.rs:42]:
```suggestion
pub fn execute_claim_mint_nft(
```

File: contracts/claim/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::ClaimMintNFT { minter_address } => {
+            try_claim_mint_nft(deps, info.sender, minter_address)
```
Comment (shanev) [inline contracts/claim/src/contract.rs:37]:
```suggestion
            execute_claim_mint_nft(deps, info.sender, minter_address)
```

File: contracts/claim/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        minter_addr,
+        &QueryMsg::MintCount {
+            address: sender.to_string(),
+        },
+    )?;
+    if count_response.count == 0 {
+        return Err(ContractError::NoMinting {});
+    }
+
+    let msg = create_claim_for_msg(sender.to_string(), sg_std::ClaimAction::MintNFT);
```
Comment (shanev) [inline contracts/claim/src/contract.rs:58]: Can import `sg_std` instead of using directly here.

File: contracts/claim/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    Ok(Response::new()
+        .add_message(msg)
+        .add_attribute("method", "claim_mint_nft")
```
Comment (shanev) [inline contracts/claim/src/contract.rs:61]:
```suggestion
        .add_attribute("action", "claim_mint_nft")
```

File: contracts/claim/src/msg.rs [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct InstantiateMsg {}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ExecuteMsg {
+    #[serde(rename = "claim_mint_nft")]
+    ClaimMintNFT { minter_address: String },
```
Comment (shanev) [inline contracts/claim/src/msg.rs:11]: Without the rename what does serde use?

File: e2e/claim_test.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/public-awesome/stargaze/v3/testutil/simapp"
+	claimtypes "github.com/public-awesome/stargaze/v3/x/claim/types"
+	"github.com/stretchr/testify/require"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+func TestClaim(t *testing.T) {
```
Comment (shanev) [inline e2e/claim_test.go:19]: Love these e2e tests!

**Review by shanev — CHANGES_REQUESTED** [review]

File: contracts/claim/src/msg.rs (hunk 7 above)
Comment (yubrew) [reply contracts/claim/src/msg.rs:11]: guessing it's claim_mint_n_f_t due to the caps

Comment (jhernandezb) [reply contracts/claim/src/msg.rs:11]: it uses `claim_mint_n_f_t`

File: contracts/claim/src/error.rs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+use cosmwasm_std::StdError;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum ContractError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("Unauthorized")]
```
Comment (yubrew) [inline contracts/claim/src/error.rs:9]: is this referenced anywhere?

File: e2e/claim_test.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+	require.NoError(t, err)
+	require.False(t, claim.ActionCompleted[claimtypes.ActionMintNFT])
+
+	app.ClaimKeeper.SetParams(ctx, claimtypes.Params{
+		AirdropEnabled:     true,
+		AirdropStartTime:   startDateTime,
+		DurationUntilDecay: claimtypes.DefaultDurationUntilDecay,
+		DurationOfDecay:    claimtypes.DefaultDurationOfDecay,
+		ClaimDenom:         claimtypes.DefaultClaimDenom,
+		AllowedClaimers: []claimtypes.ClaimAuthorization{
```
Comment (yubrew) [inline e2e/claim_test.go:206]: does this limits the contracts that minters can claim against?

File: e2e/claim_test.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+	require.Greater(t, perAction.Int64(), int64(0))
+	expectedBalance := perAction.Add(sdk.NewInt(1_900_000_000)) // user already had 19000
+	require.Equal(t,
+		expectedBalance.String(),
+		balance.Amount.String(),
+	)
+
+	claim, err = app.ClaimKeeper.GetClaimRecord(ctx, addr1)
+	require.NoError(t, err)
+	require.True(t, claim.ActionCompleted[claimtypes.ActionMintNFT])
```
Comment (yubrew) [inline e2e/claim_test.go:239]: How/where is this data stored? Does this prevents an address from double claiming?

**Review by yubrew — COMMENTED** [review]
when you create a new contract, are you using `cw-template`?

File: e2e/claim_test.go (hunk 11 above)
Comment (jhernandezb) [reply e2e/claim_test.go:239]: yup everything is taken care from the native claim module, see final test is attempting to claim again but balance doesn't move.

File: e2e/claim_test.go (hunk 10 above)
Comment (jhernandezb) [reply e2e/claim_test.go:206]: this limits an address that can submit the claim message to the chain. In this case after the contract is uploaded another gov proposal needs to be submitted with the contract address and only that instantiation can claim the airdrop.

File: contracts/claim/src/error.rs (hunk 9 above)
Comment (jhernandezb) [reply contracts/claim/src/error.rs:9]: I guess I can remove

Comment (jhernandezb) [reply contracts/claim/src/error.rs]: removed

**Review by shanev — APPROVED** [review]

**Review by yubrew — APPROVED** [review]

**Pull request merged**